src/System.Buffers.Primitives/System/Buffers/ReadOnlyBufferReader.cs

```diff
@@ -135,7 +135,7 @@ public void Advance(int byteCount)
             }
         }
 
-        internal static int Peek(BufferReader<TSequence> bytes, Span<byte> destination)
+        internal static int Peek(in BufferReader<TSequence> bytes, Span<byte> destination)
         {
             var first = bytes.UnreadSegment;
```

Comment:
Will cause copy when passed `in`? (as `BufferReader<TSequence>` is not readonly)

Should the property be expanded here to work on the fields

Reply:
Hmmm, you are right. This just moves the inefficient copies of the whole BufferReader to different places. I do not know how to avoid them with this API shape. @KrzysztofCwalina Thoughts?

Reply:
> Think the only change needed is

There are still going to be copies where Peek is called. For example, `var copied = BufferReader.Peek(reader, tempSpan);` will have a copy. You can tell that it has a copy because of if you explicitly pass it as `in reader`, it won't compile anymore.

>  C# compiler/analyser to issue a warning that its going to do a copy of a readonly variable 

Agree. It is a pit of failure that it is very easy to fall into, with `in` in particular.

Reply:
Can the parameter by by ref (as opposed to in)?

BTW, This method is a hack in general. I originally had it as an instance method but after changing the struct to be a by-ref struct, the compiler started to complain that I cannot pass stack allocated spans to it. @VSadov suggested this fix (hack?) and when I talked to @jaredpar he suggested that at some point we can add readonly methods to the language.

Reply:
When I made the parameter byref, the compiler started complain about passing spans to it.

Reply:
IIRC, `in` will only cause a copy if you don't explicitly specify.

That is `Peek(bytes, destination)` may cause a copy, but `Peek(in bytes, destination)` will elide the copy as you explicitly opted in.

@jcouv or @VSadov may remember better 😄

Reply:
> `Peek(in bytes, destination)`

The problem is this does not compile. It fails to compile with:

```
System\Buffers\BufferReader_text.cs(73,26): error CS8350: This combination of arguments to 'BufferReader.Peek<TSequence>(in BufferReader<TSequence>, Span<byte>)' is disallowed because it may expose variables referenced by parameter 'destination' outside of their declaration scope
```

Reply:
Re; warning on implicit copying. FWIW this was a behavior since v1. Invoking a regular (and so formally mutating) method on a readonly variable does not warn/error, but instead makes a copy of the receiver. 

I am not happy about it but I see how it was considered better than making things like "myStruct.ToString" a warning/error.

Reply:
Yes, 

Replacing
```C#
var first = bytes.UnreadSegment;
```
with
```C#
var first = bytes._currentSpan.Slice(bytes._index);
```
would eliminate a copy here.

Or you can make `UnreadSegment` an `in` static method...